hawkular-alerts-engine/src/main/java/org/hawkular/alerts/engine/impl/DroolsRulesEngineImpl.java

```diff
@@ -164,77 +168,115 @@ public void fire() {
 
-            long startBatching = System.currentTimeMillis();
-            for (Data data : batchData) {
-                if (null == previousData || !(data.getId().equals(previousData.getId())
-                        && data.getTenantId().equals(previousData.getTenantId()))) {
-                    kSession.insert(data);
-                    previousData = data;
+            // Keep only the least recent datum for any dataId. Remove minReportingInterval violators, defer the rest
+            Data previousData = null;
+            for (Iterator<Data> i = batchData.iterator(); i.hasNext();) {
+                Data d = i.next();
+                if (null == previousData || !(d.getId().equals(previousData.getId())
+                        && d.getTenantId().equals(previousData.getTenantId()))) {
+                    previousData = d;
 
                 } else {
-                    pendingData.add(data);
-                    log.tracef("Deferring more recent %s until older %s is processed", data, previousData);
+                    i.remove();
+                    if ((d.getTimestamp() - previousData.getTimestamp()) < minReportingInterval) {
+                        log.tracef("MinReportingInterval violation, prev: %s, removed: %s", previousData, d);
+                    } else {
+                        pendingData.add(d);
+                        log.tracef("Deferring data, keep: %s, defer: %s", previousData, d);
+                    }
                 }
             }
-            long batchingTime = System.currentTimeMillis() - startBatching;
-            log.debugf("Batching Data [%s] took [%s]", batchData.size(), batchingTime);
-            if (batchingTime > PERF_BATCHING_THRESHOLD) {
-                log.warnf("Batching Data [%s] took [%s] ms exceeding [%s] ms",
-                        batchData.size(), batchingTime, PERF_BATCHING_THRESHOLD);
-            }
 
-            if (!pendingData.isEmpty() && log.isDebugEnabled()) {
-                log.debugf("Deferring [%s] Datum(s) to next firing !!", pendingData.size());
+            if (!pendingData.isEmpty()) {
+                log.debugf("Deferring [%d] Datum(s) to next firing !!", pendingData.size());
             }
+        }
 
-            batchData.clear();
+        // insert the remaining datums into working memory
+        batchData.stream().forEach(d -> kSession.insert(d));
 
-            TreeSet<Event> batchEvent = pendingEvents;
+        long batchingTime = System.currentTimeMillis() - startBatching;
+        log.debugf("Batching Data [%s] took [%s]", batchData.size(), batchingTime);
+        if (batchingTime > PERF_BATCHING_THRESHOLD) {
+            log.warnf("Batching Data [%s] took [%s] ms exceeding [%s] ms",
+                    batchData.size(), batchingTime, PERF_BATCHING_THRESHOLD);
+        }
+    }
+
+    private void batchEvents() {
+        TreeSet<Event> batchEvents;
+        long startBatching;
+
+        synchronized (pendingEvents) {
```

Comment:
Same comment.
This should be executed from a single thread, no need to `synchronized` here.